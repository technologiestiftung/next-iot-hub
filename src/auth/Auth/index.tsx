import {
  useState,
  useEffect,
  createContext,
  FC,
  useContext,
  useCallback,
} from "react";
import { supabase } from "../supabase";
import { AuthenticatedUsersType } from "@common/types/supabase_DEPRECATED";

export const AuthProvider: FC = ({ children }) => {
  const [authenticatedUser, setUser] = useState<
    AuthContextType["authenticatedUser"]
  >();
  const [isLoadingAuth, setLoading] = useState<
    AuthContextType["isLoadingAuth"]
  >(true);
  const [accessToken, setAccessToken] = useState<
    AuthContextType["accessToken"]
  >(null);

  useEffect(() => {
    const session = supabase.auth.session();

    setUser(session?.user ?? null);
    setAccessToken(session?.access_token ?? null);
    setLoading(false);

    const { data: listener } = supabase.auth.onAuthStateChange(
      (_event, session) => {
        setUser(session?.user ?? null);
        setAccessToken(session?.access_token ?? null);
        setLoading(false);
      }
    );

    return () => {
      listener?.unsubscribe();
    };
  }, []);

  const value = {
    signIn: supabase.auth.signIn.bind(supabase.auth),
    signOut: supabase.auth.signOut.bind(supabase.auth),
    authenticatedUser: authenticatedUser ?? null,
    isLoadingAuth,
    accessToken,
  };

  return <AuthContext.Provider value={value}>{children}</AuthContext.Provider>;
};

interface SignUpDataType {
  username: string;
  email: string;
}

const signUp = async ({
  username,
  email,
}: SignUpDataType): Promise<{
  error: Error | null;
}> => {
  const myHeaders = new Headers();
  myHeaders.append("Content-Type", "application/json");

  const requestOptions = {
    method: "POST",
    headers: myHeaders,
    redirect: "follow" as const,
    body: JSON.stringify({ name: username, email }),
  };

  const baseUrl = process.env.NEXT_PUBLIC_TOKEN_API_URL || "";
  const url = `${baseUrl}/api/v3/signup`;
  try {
    const response = await fetch(url, requestOptions);
    const jsonResponse = (await response.json()) as {
      error?: string;
      message?: string;
      statusCode?: number;
    };
    if (jsonResponse.statusCode !== 200)
      return { error: new Error(jsonResponse.message) };
    return { error: null };
  } catch (e) {
    return { error: e as Error };
  }
};
interface AuthContextType {
  signIn: typeof supabase.auth.signIn;
  signOut: typeof supabase.auth.signOut;
  authenticatedUser: AuthenticatedUsersType | null;
  isLoadingAuth: boolean;
  accessToken: string | null;
}

const defaultValue = {
  signIn: supabase.auth.signIn.bind(supabase.auth),
  signUp: supabase.auth.signIn.bind(supabase.auth),
  signOut: supabase.auth.signOut.bind(supabase.auth),
  authenticatedUser: null,
  isLoadingAuth: true,
  accessToken: null,
};

const AuthContext = createContext<AuthContextType>(defaultValue);

interface AuthHookReturnType extends Omit<AuthContextType, "signIn"> {
  magicLinkWasSent: boolean;
  isAuthenticating: boolean;
  error: string | null;
  signUp: (data: SignUpDataType) => Promise<void>;
  signIn: (data: { email: string }) => Promise<void>;
}

export const useAuth = (): AuthHookReturnType => {
  const authContext = useContext(AuthContext);
  const { signIn, ...auth } = authContext;
  const [magicLinkWasSent, setMagicLinkWasSent] = useState<boolean>(false);
  const [isAuthenticating, setIsAuthenticating] = useState<boolean>(false);
  const [error, setError] = useState<string | null>(null);

  const signInHandler = useCallback(
    async (data: { email: string }): Promise<void> => {
      setIsAuthenticating(true);
      const { error } = await signIn(data);

      if (error) setError(error.message);
      if (!error) setMagicLinkWasSent(true);
      setIsAuthenticating(false);
    },
    [signIn, setMagicLinkWasSent, setIsAuthenticating]
  );

  const signUpHandler = useCallback(
    async (data: { username: string; email: string }): Promise<void> => {
      setIsAuthenticating(true);
      const { error } = await signUp(data);

      if (error) setError(error.message);
      if (!error) setMagicLinkWasSent(true);
      setIsAuthenticating(false);
    },
    [setMagicLinkWasSent, setIsAuthenticating]
  );
  return {
    ...auth,
    error,
    magicLinkWasSent,
    isAuthenticating,
    signIn: signInHandler,
    signUp: signUpHandler,
  };
};
