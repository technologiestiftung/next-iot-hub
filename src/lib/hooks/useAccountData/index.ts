import useSWR from "swr";
import {
  AccountWithSensorsType,
  getAccountDataByUsername,
} from "@lib/requests/getAccountDataByUsername";

interface useAccountDataParamsType {
  username?: string;
  initialData?: AccountWithSensorsType;
}

interface useAccountDataReturnType {
  isLoading: boolean;
  account: AccountWithSensorsType | null;
  error: Error | null;
}

export const useAccountData = ({
  username,
  initialData,
}: useAccountDataParamsType): useAccountDataReturnType => {
  const params = [`account-${username || "no"}-data`];
  const { data, error } = useSWR<AccountWithSensorsType | undefined, Error>(
    params,
    () => getAccountDataByUsername(username),
    { initialData }
  );

  return {
    isLoading: !data && !error,
    account: data || null,
    error: error || null,
  };
};
