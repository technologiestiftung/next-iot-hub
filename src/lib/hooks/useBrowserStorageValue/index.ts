import { useCallback, useEffect, useState } from "react";

function useBrowserStorageValue<ValueType = unknown>(
  key: string,
  initialValue: ValueType | null,
  storageType: "session" | "local"
): [value: ValueType | null, setValue: (val: ValueType | null) => void] {
  const [value, setValue] = useState<ValueType | null>(initialValue);

  useEffect(() => {
    const storage = storageType === "local" ? localStorage : sessionStorage;
    if (!storage) return;
    const rawVal = storage.getItem(key);
    if (!rawVal) return;
    try {
      const parsedVal = JSON.parse(rawVal) as ValueType;
      setValue(parsedVal);
    } catch (error) {
      console.error(error);
    }
  }, []);

  const setter = useCallback(
    (val: ValueType | null) => {
      const storage = storageType === "local" ? localStorage : sessionStorage;
      if (!storage || !val) return;
      storage.setItem(key, JSON.stringify(val));
      setValue(val);
    },
    [key, setValue, storageType]
  );

  return [value, setter];
}

export function useLocalStorageValue<ValueType = unknown>(
  key: string,
  initialValue: ValueType | null
): [value: ValueType | null, setValue: (val: ValueType | null) => void] {
  return useBrowserStorageValue<ValueType>(key, initialValue, "local");
}

export function useSessionStorageValue<ValueType = unknown>(
  key: string,
  initialValue: ValueType | null
): [value: ValueType | null, setValue: (val: ValueType | null) => void] {
  return useBrowserStorageValue<ValueType>(key, initialValue, "session");
}
