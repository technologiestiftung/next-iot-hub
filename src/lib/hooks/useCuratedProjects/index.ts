import {
  mapPublicProject,
  projectQueryString,
  PublicProject,
} from "./../usePublicProjects/index";
import { supabase } from "@auth/supabase";
import useSWR from "swr";

import { ProjectsType } from "@common/types/supabase_DEPRECATED";

const RECORDS_LIMIT = 500;
const PROJECTS_LIMIT = 3;

export const getCuratedProjects = async (): Promise<PublicProject[]> => {
  const { data, error } = await supabase
    .from<ProjectsType>("projects")
    .select(projectQueryString)
    //FIXME: the typecasting
    .order("recordedAt" as keyof ProjectsType, {
      foreignTable: "devices.records",
      ascending: false,
    })
    .limit(PROJECTS_LIMIT)
    .limit(RECORDS_LIMIT, { foreignTable: "devices.records" });

  if (error) throw error;
  if (!data) return [];
  const projects = data?.map(mapPublicProject);

  return projects;
};

export const useCuratedProjects = (): {
  data: PublicProject[] | null;
  error: Error | null;
} => {
  const { data, error } = useSWR<PublicProject[], Error>(
    "useCuratedProjects",
    getCuratedProjects
  );

  return {
    data: data || null,
    error: error || null,
  };
};
