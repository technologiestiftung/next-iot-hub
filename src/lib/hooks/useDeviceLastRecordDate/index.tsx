import { getDeviceLastRecordDate } from "@lib/requests/getDeviceLastRecordDate";
import useSWR from "swr";

interface UseDeviceLastRecordDateReturnType {
  isLoading: boolean;
  lastRecordDate: string | undefined;
  error: Error | null;
}

const fetchDeviceLastRecordDate = async (
  deviceId: number | undefined
): Promise<string | null> => {
  if (!deviceId) return null;

  const lastRecordDate = await getDeviceLastRecordDate(deviceId);
  return lastRecordDate || null;
};

export const useDeviceLastRecordDate = (
  deviceId: number | undefined
): UseDeviceLastRecordDateReturnType => {
  const params = [`device-${deviceId || "no"}-lastRecordDate`, deviceId];
  const { data: lastRecordDate, error } = useSWR<string | null, Error>(
    params,
    () => fetchDeviceLastRecordDate(deviceId)
  );

  return {
    isLoading: lastRecordDate === null,
    lastRecordDate: lastRecordDate || undefined,
    error: error || null,
  };
};
