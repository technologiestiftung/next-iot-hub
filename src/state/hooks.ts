import { createTypedHooks } from "easy-peasy";
import { StoreModel } from "./model";

const {
  useStoreActions,
  useStoreState,
  useStoreDispatch,
  useStore,
} = createTypedHooks<StoreModel>();

export { useStoreActions, useStoreState, useStoreDispatch, useStore };
