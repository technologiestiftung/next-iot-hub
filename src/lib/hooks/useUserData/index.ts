import useSWR from "swr";
import { supabase } from "@auth/supabase";
import {
  UsersType,
  AuthenticatedUsersType,
  ProjectsType,
  DevicesType,
} from "@common/types/supabase";
import { useAuth } from "@auth/Auth";

interface UserDataType {
  user: UsersType;
  projects: ProjectsType[];
}

const fetchUserData = async (
  authenticatedUser: AuthenticatedUsersType | null
): Promise<UserDataType> => {
  if (!authenticatedUser) throw new Error("Not authenticated");

  const { data: user, error: user_error } = await supabase
    .from<UsersType>("users")
    .select(
      `
      name
    `
    )
    .eq("id", authenticatedUser.id)
    .single();

  if (user_error) throw user_error;

  const { data: projects, error: projects_error } = await supabase
    .from<ProjectsType>("projects")
    .select(
      `
      id,
      name,
      description,
      connectype,
      location,
      category:categoryId (
        id,
        name,
        description
      ),
      devices (
        id,
        externalId,
        name
      )
    `
    )
    .eq("userId", authenticatedUser.id);

  if (projects_error) throw projects_error;
  else if (!user)
    throw new Error(`User with id "${authenticatedUser.id} was not found"`);
  else if (!projects)
    throw new Error(
      `Projects for user with id "${authenticatedUser.id} was not found"`
    );
  else return { user, projects };
};

const addDevice = async (
  device: Omit<DevicesType, "id">
): Promise<DevicesType[] | null | Error> => {
  if (!device.userId) throw new Error("Not authenticated");
  const { data, error } = await supabase
    .from<DevicesType>("devices")
    .insert([device]);
  if (error) throw error;
  else return data;
};

const updateDevice = async ({
  id,
  userId,
  externalId,
  name,
}: Partial<DevicesType>): Promise<DevicesType[] | null | Error> => {
  if (!userId) throw new Error("Not authenticated");
  const { data, error } = await supabase
    .from<DevicesType>("devices")
    .update({ externalId, name })
    .eq("id", id)
    .eq("userId", userId);
  if (error) throw error;
  else return data;
};

const deleteDevice = async (
  id: number,
  userId: string | undefined
): Promise<DevicesType[] | null | Error> => {
  if (!userId) throw new Error("Not authenticated");
  const { data, error } = await supabase
    .from<DevicesType>("devices")
    .delete()
    .eq("id", id)
    .eq("userId", userId);
  if (error) throw error;
  else return data;
};

const addProject = async (
  project: Omit<ProjectsType, "id" | "createdAt">
): Promise<ProjectsType[] | null | Error> => {
  if (!project.userId) throw new Error("Not authenticated");
  const { data, error } = await supabase
    .from<ProjectsType>("projects")
    .insert([project]);
  if (error) throw error;
  else return data;
};

const updateProject = async ({
  id,
  name,
  description,
  connectype,
  location,
  userId,
  categoryId,
}: Partial<ProjectsType>): Promise<ProjectsType[] | null | Error> => {
  if (!userId) throw new Error("Not authenticated");
  const { data, error } = await supabase
    .from<ProjectsType>("projects")
    .update({ name, description, connectype, location, categoryId })
    .eq("id", id)
    .eq("userId", userId);
  if (error) throw error;
  else return data;
};

const deleteProject = async (
  id: number,
  userId: string | undefined
): Promise<ProjectsType[] | null | Error> => {
  if (!userId) throw new Error("Not authenticated");
  const { data, error } = await supabase
    .from<ProjectsType>("projects")
    .delete()
    .eq("id", id)
    .eq("userId", userId);
  if (error) throw error;
  else return data;
};

const updateName = async (
  name: string,
  userId: string | undefined
): Promise<UsersType[] | null | Error> => {
  if (!userId) throw new Error("Not authenticated");
  const { data, error } = await supabase
    .from<UsersType>("users")
    .update({ name })
    .eq("id", userId);
  if (error) throw error;
  else return data;
};
const updateEmail = async (
  email: string,
  userId: string | undefined
): Promise<AuthenticatedUsersType | null | Error> => {
  if (!userId) throw new Error("Not authenticated");
  const { user, error } = await supabase.auth.update({
    data: { email },
  });
  if (error) throw error;
  else return user as AuthenticatedUsersType;
};
const deleteUser = async (
  userId: string | undefined
): Promise<any[] | null | Error> => {
  if (!userId) throw new Error("Not authenticated");
  const { data, error } = await supabase.rpc("delete_user");
  if (error) throw error;
  else return data;
};

export const useUserData = (): {
  authenticatedUser: AuthenticatedUsersType | null;
  user: UsersType | null;
  projects: ProjectsType[] | null;
  error: Error | null;
  addDevice: ({
    externalId,
    name,
    projectId,
  }: DevicesType) => Promise<DevicesType[] | null | Error>;
  updateDevice: ({
    id,
    externalId,
    name,
  }: DevicesType) => Promise<DevicesType[] | null | Error>;
  deleteDevice: (id: number) => Promise<DevicesType[] | null | Error>;
  addProject: ({
    name,
    description,
    connectype,
    location,
    categoryId,
  }: ProjectsType) => Promise<ProjectsType[] | null | Error>;
  updateProject: ({
    id,
    name,
    description,
    connectype,
    location,
    categoryId,
  }: ProjectsType) => Promise<ProjectsType[] | null | Error>;
  deleteProject: (id: number) => Promise<ProjectsType[] | null | Error>;
  updateName: (name: string) => Promise<UsersType[] | null | Error>;
  updateEmail: (
    email: string
  ) => Promise<AuthenticatedUsersType | null | Error>;
  deleteUser: () => Promise<any[] | null | Error>;
} => {
  const { authenticatedUser } = useAuth();
  const { data, error } = useSWR<UserDataType, Error>(
    ["user", authenticatedUser?.id],
    () => fetchUserData(authenticatedUser)
  );

  return {
    authenticatedUser: authenticatedUser || null,
    user: data?.user || null,
    projects: data?.projects || null,
    error: error || null,
    addDevice: ({ externalId, name, projectId }) =>
      addDevice({ externalId, name, projectId, userId: authenticatedUser?.id }),
    updateDevice: ({ id, externalId, name }) =>
      updateDevice({ id, externalId, name, userId: authenticatedUser?.id }),
    deleteDevice: id => deleteDevice(id, authenticatedUser?.id),
    addProject: ({ name, description, connectype, location, categoryId }) =>
      addProject({
        name,
        description,
        connectype,
        location,
        categoryId,
        userId: authenticatedUser?.id,
      }),
    updateProject: ({
      id,
      name,
      description,
      connectype,
      location,
      categoryId,
    }) =>
      updateProject({
        id,
        name,
        description,
        connectype,
        location,
        categoryId,
        userId: authenticatedUser?.id,
      }),
    deleteProject: id => deleteProject(id, authenticatedUser?.id),
    updateName: (name: string) => updateName(name, authenticatedUser?.id),
    updateEmail: (email: string) => updateEmail(email, authenticatedUser?.id),
    deleteUser: () => deleteUser(authenticatedUser?.id),
  };
};
