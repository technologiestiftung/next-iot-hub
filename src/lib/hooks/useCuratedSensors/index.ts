import useSWR from "swr";
import { ParsedSensorType } from "@lib/hooks/usePublicSensors";
import { getPublicSensors } from "@lib/requests/getPublicSensors";
import { getSensorsRecords } from "@lib/requests/getSensorsRecords";

// The following is a quickfix because we currently know what our curated sensors are
// In the future we might wanna save them in a DB table
// TODO: Refactor this!
const isProduction =
  process.env["NODE_ENV"] === "production" &&
  process.env["VERCEL_ENV"] !== "preview";
const SENSORS_IDS = [22, 23, 24, 25, 27, 28, 29];

export const getCuratedSensorImageById = (id: number): string => {
  if (isProduction) return `/images/sensors/${id}.jpeg`;
  const indexOfId = SENSORS_IDS.indexOf(id);
  return `/images/sensors/${SENSORS_IDS[indexOfId]}.jpeg`;
};

export const getCuratedSensors = async (): Promise<ParsedSensorType[]> => {
  const { sensors } = await getPublicSensors();
  const sensorsRecordsMap = await getSensorsRecords(
    sensors.map(({ id }) => id)
  );
  const curated = sensors
    .filter(sensor => SENSORS_IDS.includes(sensor.id))
    .map(sensor => ({
      ...sensor,
      parsedRecords: (sensorsRecordsMap[sensor.id] || []).slice(-500),
    }));

  return curated;
};

export const useCuratedSensors = (): {
  data: ParsedSensorType[] | null;
  error: Error | null;
} => {
  const { data, error } = useSWR<ParsedSensorType[], Error>(
    "useCuratedSensors",
    getCuratedSensors
  );

  return {
    data: error ? null : data || null,
    error: error || null,
  };
};
