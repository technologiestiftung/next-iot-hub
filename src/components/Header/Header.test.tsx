import { AuthProvider } from "@auth/Auth";
import * as userDataHook from "@lib/hooks/useUserData";
import { render, screen, waitFor } from "@testing-library/react";
import { Header } from ".";

const originalUserDataHook = userDataHook.useUserData;

describe("Header component", () => {
  beforeEach(() => {
    // Ignored because of the reassignment for mock purposes
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    userDataHook.useUserData = jest.fn().mockReturnValue({
      user: { name: "JohnDoe" },
    });
  });

  afterEach(() => {
    // Ignored because of the reassignment for mock purposes
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    userDataHook.useUserData = originalUserDataHook;
  });
  it("should render the Stadtpuls logo", async () => {
    render(<Header />);
    const logo = screen.getByRole("img");
    await waitFor(() => expect(logo).toBeInTheDocument());
  });
  it("should render the tsb logo", async () => {
    render(<Header />);
    const logo = document.querySelector("svg");
    await waitFor(() => expect(logo).toBeInTheDocument());
  });
  it("should render the 'authentication' link if logged out", async () => {
    // Ignored because of the reassignment for mock purposes
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    userDataHook.useUserData = jest.fn().mockReturnValue({
      user: null,
    });
    render(<Header />);
    const link = screen.getByText(/Anmeldung/g);
    await waitFor(() => expect(link).toBeInTheDocument());
  });
  it("should render the 'username' link if logged in", async () => {
    render(
      <AuthProvider>
        <Header />
      </AuthProvider>
    );
    const username = screen.getByText("JohnDoe");

    await waitFor(() => {
      expect(username).toBeInTheDocument();
    });
  });
});
