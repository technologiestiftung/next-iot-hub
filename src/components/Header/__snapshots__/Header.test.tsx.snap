// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots Footer Default 1`] = `
<footer
  className="css-605ip6-Footer"
>
  <div
    className="css-177thsy"
  >
    <div
      className="css-4cffwv"
    >
      <div
        className="css-1tospi0-Footer"
      >
        <div
          className="css-vurnku"
        >
          Ein Projekt der:
        </div>
      </div>
      <img
        alt="Logo der Technologiestiftung Berlin"
        className="css-1t6cmvr-Footer"
        src="/images/tsb-logo.svg"
      />
    </div>
    <div
      className="css-4cffwv"
    >
      <div
        className="css-1tospi0-Footer"
      >
        <div
          className="css-vurnku"
        >
          Gefördert von:
        </div>
      </div>
      <img
        alt="Logo der Senatsverwaltung für Wirtschaft, Energie und Betriebe"
        className="css-h1k1oh-Footer"
        src="/images/senweb-logo.svg"
      />
    </div>
  </div>
  <div
    className="css-18f9hr4-Footer"
  >
    <a
      className="css-ta8zj8"
      href="https://www.technologiestiftung-berlin.de/de/impressum/"
      rel="noopener noreferrer"
      target="_blank"
    >
      Impressum
    </a>
    <a
      className="css-1cve2jh"
      href="https://www.technologiestiftung-berlin.de/de/datenschutz/"
      rel="noopener noreferrer"
      target="_blank"
    >
      Datenschutz
    </a>
  </div>
</footer>
`;

exports[`Storyshots Header Default 1`] = `
<header
  className="css-4werlw-Header"
>
  <a
    className="css-1kxglq2-Header"
    onClick={[Function]}
    onMouseEnter={[Function]}
  >
    <div
      className="css-4cffwv"
    >
      <img
        alt="Logo des Berlin IoT Hub"
        className="css-k3qijy-Header"
        src="/images/datahub-logo.svg"
      />
      <h2
        className="css-12hmaxm-Header"
      >
        Berlin 
        <span
          className="css-u111td-Header"
        >
          IoT Hub
        </span>
      </h2>
    </div>
  </a>
</header>
`;
